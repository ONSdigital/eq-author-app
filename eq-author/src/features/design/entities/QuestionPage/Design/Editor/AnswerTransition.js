import { CSSTransition } from "react-transition-group";
import PropTypes from "prop-types";
import styled from "styled-components";

const timeout = props => props.timeout;
const halfTimeout = props => props.timeout / 2;

const handleExit = node => {
  const { height } = node.getBoundingClientRect();
  node.style.height = `${height}px`;
};

const AnswerTransition = styled(CSSTransition).attrs({
  classNames: "answer",
  onExit: () => handleExit
})`
  position: relative;

  &.answer-enter {
    opacity: 0;
    transform: scale(0.9);
    z-index: 200;
  }

  &.answer-enter-active {
    opacity: 1;
    transform: scale(1);
    transition: opacity ${timeout}ms ease-out,
      transform ${timeout}ms cubic-bezier(0.175, 0.885, 0.32, 1.275);
  }

  &.answer-exit {
    opacity: 1;
    transform: scale(1);
  }

  &.answer-exit-active {
    opacity: 0;
    height: 0 !important;
    transform: scale(0.9);
    transition: opacity ${halfTimeout}ms ease-out,
      height ${halfTimeout}ms ease-in ${halfTimeout}ms,
      transform ${halfTimeout}ms ease-in;
  }
`;

AnswerTransition.propTypes = {
  timeout: PropTypes.number,
  children: PropTypes.element
};

AnswerTransition.defaultProps = {
  timeout: 200
};

export default AnswerTransition;
