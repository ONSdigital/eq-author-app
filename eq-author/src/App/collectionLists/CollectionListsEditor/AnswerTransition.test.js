import React from "react";
import { shallow, mount } from "enzyme";

import AnswerTransition from "./AnswerTransition";

describe("components/QuestionPageEditor/AnswerTransition", () => {
  it("should render", () => {
    expect(
      shallow(
        <AnswerTransition>
          <div>Content</div>
        </AnswerTransition>
      )
    ).toMatchSnapshot();
  });

  it("should set the final height of the element to the height of the element at the end of transition", () => {
    const wrapper = mount(
      <AnswerTransition>
        <div>Content</div>
      </AnswerTransition>
    );
    const node = {
      getBoundingClientRect: jest.fn().mockReturnValue({ height: 10 }),
      style: {
        height: null,
      },
    };

    wrapper.find("CSSTransition").prop("onExit")()(node);
    expect(node.getBoundingClientRect).toHaveBeenCalled();
    expect(node.style.height).toEqual("10px");
  });
});
