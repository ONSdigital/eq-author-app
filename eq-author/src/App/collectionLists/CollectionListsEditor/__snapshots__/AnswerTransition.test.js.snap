// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/QuestionPageEditor/AnswerTransition should render 1`] = `
<AnswerTransition__AnswerTransitionWrapper
  timeout={200}
>
  <div>
    <div>
      Content
    </div>
  </div>
</AnswerTransition__AnswerTransitionWrapper>
`;
